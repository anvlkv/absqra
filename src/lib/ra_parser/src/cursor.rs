use ra_lexer::token::{Token};
use ra_lexer::cursor::Position;
use core::iter::Peekable;

pub(crate) struct Cursor<'token, I> 
    where I: Iterator<Item = Token<'token>> 
{
    pub position: Position,
    pub level: u16,
    tokens: Peekable<I>,
    consumed_len: usize
}

impl <'token, I> Cursor<'token, I> 
where I: Iterator<Item = Token<'token>> 
{
    pub fn new(tokens: I) -> Self
        
    {
        Self {
            position: Position(0, 0),
            level: 0,
            tokens: tokens.peekable(),
            consumed_len: 0
        }
    }

    pub fn first_ahead(&mut self) -> Option<&Token<'token>> {
        self.tokens.peek()
    }

    pub fn bump(&mut self) -> Option<Token<'token>> {
        let tok = self.tokens.next();
        match tok {
            Some(token) => {
                self.position = token.position.1;
                self.level = token.level;
                self.consumed_len += 1;
                Some(token)
            }
            None => None
        }
    }

    pub fn is_eof(& mut self) -> bool {
        self.first_ahead() == None
    }

    // pub fn do_while<F, D>(&mut self, mut test: F, mut cb: D)
    //     where F: FnMut (u16, Position, TokenKind) -> bool,
    //     D: FnMut (Token)
    // {
    //     loop {
    //         let next_token = self.first_ahead();
    //         match next_token {
    //             Some(t) => {
    //                 if test(t.level, t.position.0, t.kind) {
    //                     cb(self.bump().unwrap())
    //                 }
    //                 else {
    //                     break
    //                 }
    //             },
    //             None => break
    //         }
    //     }
    // }

    // pub fn read_within(&mut self, open: TokenKind, close: TokenKind) -> Vec<Token> {
    //     let mut read: Vec<Token> = Vec::new();

    //     loop {
    //         let next_token = self.first_ahead();

    //         match next_token {
    //             Some(tok) => {
    //                 match tok.kind {
    //                     k if k == open => {
    //                         let open_token = self.bump().unwrap();
    //                         let mut nested = self.read_within(open_token.kind, close.clone());
    //                         read.append(&mut nested);
    //                     },
    //                     k if k == close => {
    //                         self.bump();
    //                         break
    //                     },
    //                     _ => read.push(self.bump().unwrap())
    //                 }
    //             },
    //             None => break
    //         }

    //         // println!("{:?}", &read);
    //     }

    //     read
    // }
}