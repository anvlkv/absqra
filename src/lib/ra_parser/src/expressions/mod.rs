pub (crate) mod buffered;


#[macro_use]
mod repeat_kindly_macros;

pub mod output_expression_kind;
pub mod output_expression;
pub mod input_expression;
pub mod reference_expression;
pub mod annotation_expression;
pub mod context_expression;
pub mod content_expression;
pub mod grouping_expression;
pub mod logic_operation;
pub mod logic_expression;
pub mod operation_expression;
pub mod procedure_expression;
pub mod math_operation;
pub mod expression;



use super::errors::ParserError;
use super::parsed_by_token::ParsedByToken;
use ra_lexer::token::{RaToken, TokenKind};
use failure::Backtrace;
