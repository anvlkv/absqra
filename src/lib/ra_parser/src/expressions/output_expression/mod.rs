use super::*;

pub mod output_expression;
pub mod expression_member;


use expression_member::OutputExpressionMember;
pub use output_expression::OutputExpression;