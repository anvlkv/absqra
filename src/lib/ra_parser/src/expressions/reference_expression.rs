
use super::traits::{*};
use super::errors::ParserError;

use ra_lexer::cursor::Position;
use ra_lexer::token::{Token, TokenKind};

#[derive(Clone, Debug, PartialEq)]
pub struct ReferenceExpression<'a> (Token<'a>, Option<Box<ReferenceExpression<'a>>>);

impl<'a> ReferenceExpression<'a> {
    pub fn new(token: Token<'a>) -> Result<Self, ParserError<'a>> {
        match token.kind.unwrap() {
            TokenKind::Identifier(_) => Ok(Self(token, None)),
            _ => Err(ParserError::ExpectedAGotB(token, vec![TokenKind::Identifier("")]))
        }
    }
}

impl<'a> Leveled for ReferenceExpression<'a> {
    fn get_level(&self) -> u16 {
        let ReferenceExpression(first_token, _) = self;
        first_token.level
    }
}

impl<'a> Positioned for ReferenceExpression<'a> {
    fn get_position(&self) -> (Position, Position) {
        let ReferenceExpression(first_token, next) = self;

        let start_position = first_token.position.0;

        let end_position = {
            if next.is_some() {
                next.as_ref().unwrap().get_position().1
            }
            else {
                first_token.position.1
            }
        };

        (start_position, end_position)
    }
}

impl<'a> Expandable<'a, ReferenceExpression<'a>, Token<'a>> for ReferenceExpression<'a> {
    fn append_item(self, token: Token<'a>) -> Result<ReferenceExpression<'a>, ParserError<'a>> {
        let ReferenceExpression(first_token, next) = self;
        if next.is_none() {
            Ok(ReferenceExpression(first_token, Some(Box::new(ReferenceExpression::new(token)?))))
        }
        else {
            let updated_expression = next.unwrap().append_item(token)?;
            Ok(ReferenceExpression(first_token, Some(Box::new(updated_expression))))
        }
    }
}