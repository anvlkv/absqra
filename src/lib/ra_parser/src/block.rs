


use super::expressions::output_expression::OutputExpression;
use super::expressions::context_expression::ContextExpression;
use super::expressions::input_expression::InputExpression;
use super::expressions::reference_expression::ReferenceExpression;
use super::expressions::annotation_expression::AnnotationExpression;
use super::expressions::content::Content;
use super::expressions::traits::{*};

use super::errors::ParserError;

use ra_lexer::token::{Token, TokenKind};
use ra_lexer::cursor::Position;

#[derive(Clone, Debug, PartialEq)]
pub enum BlockKind<'a> {
    Program,
    Output(OutputExpression<'a>),
    Input(bool, Option<InputExpression<'a>>),
    Declaration(Option<Token<'a>>),
    Invocation(Option<Token<'a>>, Option<InputExpression<'a>>),
    Reference(Option<ReferenceExpression<'a>>),
    ContextModification(Option<ContextExpression>),
    Content(Content<'a>),
    Annotation(Option<AnnotationExpression<'a>>),
    Union(usize)
}

impl<'a> Default for BlockKind<'a> {
    fn default() -> Self {
        Self::Program
    }
}

#[derive(Clone, Debug, PartialEq, Default)]
pub struct Block<'a> {
    pub kind: BlockKind<'a>,
    pub children: Vec<Block<'a>>,
    first_token: Token<'a>,
}

impl <'a> Block<'a> {
    pub fn new(first_token: Token<'a>) -> Result<Self, ParserError<'a>> {
        let kind = Self::parse_block_kind(first_token.clone())?;

        Ok(Self {
            kind,
            first_token,
            children: Vec::new()
        })
    }

    fn parse_block_kind(token: Token<'a>) -> Result<BlockKind, ParserError<'a>> {
        if token.kind.is_none() {
            return  Ok(BlockKind::Program)
        }
        
        let token_kind = token.kind.unwrap();

        match token_kind {
            TokenKind::Identifier(_)
            | TokenKind::Float(_)
            | TokenKind::Int(_)
            | TokenKind::OpenParentheses => Ok(BlockKind::Output(OutputExpression::new(token)?)),
            TokenKind::Exclamation => Ok(BlockKind::Invocation(None, None)),
            TokenKind::Plus | TokenKind::Greater => Ok(BlockKind::Input(token_kind == TokenKind::Plus, None)),
            TokenKind::ContentBlock => Ok(BlockKind::Content(Content::new(token)?)),
            TokenKind::At => Ok(BlockKind::Reference(None)),
            TokenKind::Colon => Ok(BlockKind::Declaration(None)),
            TokenKind::OpenCurlyBrace => Ok(BlockKind::ContextModification(None)),
            TokenKind::HashPound => Ok(BlockKind::Annotation(None)),
            _ => Err(ParserError::ExpectedAGotB(token, vec![
                TokenKind::Identifier(""), 
                TokenKind::Float(0.0), 
                TokenKind::Int(0), 
                TokenKind::OpenParentheses, 
                TokenKind::Exclamation, 
                TokenKind::Plus, 
                TokenKind::Greater, 
                TokenKind::ContentBlock, 
                TokenKind::At, 
                TokenKind::Colon, 
                TokenKind::OpenCurlyBrace, 
                TokenKind::HashPound
            ]))
        }
    }
}

impl<'a> Leveled for Block<'a> {
    fn get_level(&self) -> u16{
        match self.kind {
            BlockKind::Program => 0,
            _=> self.first_token.level
        }
    }
}

impl<'a> Positioned for Block<'a> {
    fn get_position(&self) -> (Position, Position) {
        // let first_child = self.children.first();
        let last_child = self.children.last();
        let start_position = self.first_token.position.0;
        let end_position = {
            if last_child.is_some() {
                last_child.unwrap().get_position().1
            }
            else {
                let expression_position = match &self.kind {
                    BlockKind::Output(expression) => Some(expression.get_position().1),
                    BlockKind::Input(_, expression) => {
                        if expression.is_some() {
                            Some(expression.as_ref().unwrap().get_position().1)
                        }
                        else {
                            None
                        }
                    },
                    BlockKind::Invocation(token, expression) => {
                        if expression.is_some() {
                            Some(expression.as_ref().unwrap().get_position().1)
                        }
                        else if token.is_some() {
                            Some(token.unwrap().position.1)
                        }
                        else {
                            None
                        }
                    },
                    BlockKind::Declaration(declared_token) => {
                        if declared_token.is_some() {
                            Some(declared_token.unwrap().position.1)
                        }
                        else {
                            None
                        }
                    },
                    BlockKind::Annotation(expression) => {
                        if expression.is_some() {
                            Some(expression.as_ref().unwrap().get_position().1)
                        }
                        else {
                            None
                        }
                    },
                    BlockKind::Reference(expression) => {
                        if expression.is_some() {
                            Some(expression.as_ref().unwrap().get_position().1)
                        }
                        else {
                            None
                        }
                    },
                    _ => None
                };

                if expression_position.is_some() {
                    expression_position.unwrap()
                }
                else {
                    self.first_token.position.1
                }
            }
        };

        (start_position, end_position)
    }
}

impl<'a> Expandable<'a, Block<'a>, Token<'a>> for Block<'a> {
    fn append_item(self, token: Token<'a>) -> Result<Block<'a>, ParserError<'a>> {
        let mut block = self.clone();
        match self.kind {
            BlockKind::Program 
            | BlockKind::Union(_) 
            | BlockKind::Content(_) => Err(ParserError::InvalidBlock),
            BlockKind::Output(expression) => {
                let updated_expression = expression.append_item(token)?;
                block.kind = BlockKind::Output(updated_expression);
                Ok(block)
            },
            BlockKind::Input(multiple, expression) => {
                let new_expression;
                if expression.is_some() {
                    new_expression = expression.unwrap().append_item(token, None)?;
                }
                else {
                    new_expression = InputExpression::new(token)?;
                }

                block.kind = BlockKind::Input(multiple, Some(new_expression));
                Ok(block)
            },
            BlockKind::Annotation(expression) => {
                let new_expression;
                if expression.is_some() {
                    // let AnnotationExpression(first_token, next) = expression.clone().unwrap();

                    // TODO: should append to existing expression or push to block children
                    // TODO: what are possible annotated blocks

                    new_expression = expression.unwrap().append_item(token)?;
                }
                else {
                    new_expression = AnnotationExpression::new(token)?;
                }

                block.kind = BlockKind::Annotation(Some(new_expression));
                Ok(block)
            },
            BlockKind::Declaration(declared_token) => {
                if declared_token.is_some() {
                    Err(ParserError::InvalidBlock)
                }
                else {
                    block.kind = BlockKind::Declaration(Some(token));
                    Ok(block)
                }
            },
            BlockKind::Invocation(invoked_token, expression) => {
                if invoked_token.is_none() {
                    match token.kind.unwrap() {
                        TokenKind::Identifier(_) => {
                            block.kind = BlockKind::Invocation(Some(token), None);
                        },
                        _ => return Err(ParserError::ExpectedAGotB(token, vec![TokenKind::Identifier("")]))
                    }
                }
                else if expression.is_none() {
                    block.kind = BlockKind::Invocation(invoked_token, Some(InputExpression::new(token)?));
                }
                else {
                    block.kind = BlockKind::Invocation(invoked_token, Some(expression.unwrap().append_item(token, None)?))
                }

                Ok(block)
            },
            BlockKind::Reference(expression) => {
                let new_expression;
                if expression.is_some() {
                    new_expression = expression.unwrap().append_item(token)?;
                }
                else {
                    new_expression = ReferenceExpression::new(token)?;
                }

                block.kind = BlockKind::Reference(Some(new_expression));
                Ok(block)
            },
            BlockKind::ContextModification(expression) => {
                let new_expression;
                if expression.is_some() {
                    new_expression = expression.unwrap().append_item(token)?;
                }
                else {
                    new_expression = ContextExpression::new(token)?;
                }

                block.kind = BlockKind::ContextModification(Some(new_expression));
                Ok(block)
            }
        }
    }
}