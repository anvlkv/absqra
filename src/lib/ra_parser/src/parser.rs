use crate::errors::ParserError;
use crate::block::Block;
use crate::parsed_by_token::ParsedByToken;
use ra_lexer::tokenize;
use ra_lexer::token::TokenKind;

pub fn parse<'a>(input: &'a str) -> Result<Block, (Vec<ParserError>, Option<Block>)> {
    let mut errors: Vec<ParserError> = Vec::new();
    let mut tokens_stream = tokenize(input);
    
    let mut block = {
        match Block::new_program() {
            Ok(b) => b,
            Err(e) => {
                errors.extend(e);
                return Err((errors, None));
            }
        }
    };

    while let Some(token_result) = tokens_stream.next() {
        match token_result {
            Ok(token) => {
                if token.kind == TokenKind::Comment {
                    continue;
                }
                else {
                    match block.clone().append_token(token) {
                        Ok(blk) => block=blk,
                        Err(e) => errors.extend(e)
                    }
                }
            },
            Err(e) => {
                errors.push(e.into());
            }
        }
    }

    if errors.len() == 0 {
        Ok(*block)
    }
    else {
        Err((errors, Some(*block)))
    }
}