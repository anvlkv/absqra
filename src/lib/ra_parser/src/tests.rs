mod lib {
    use ra_lexer::cursor::Position;
    use ra_lexer::token::{Token, TokenKind};

    use crate::block::BlockKind;
    use crate::expressions::output_expression::ExpressionMember;
    use crate::expressions::output_expression::{MathOperation, OperationKind, OutputExpression};
    use crate::expressions::reference_expression::ReferenceExpression;
    use crate::parser::parse;

    #[test]
    fn it_should_return_program_block() {
        assert_eq!(parse("abc").expect("can't parse").kind, BlockKind::Program);
    }

    #[test]
    fn it_should_parse_block_level_expressions() {
        let program = parse("abc + 2").expect("can't parse");
        let parsed = program.children.iter().nth(0).unwrap();
        assert_eq!(
            parsed.kind,
            BlockKind::Output(OutputExpression(
                Box::new(ExpressionMember::ReferenceExpression(ReferenceExpression(Token {
                    kind: Some(TokenKind::Identifier("abc")),
                    position: (Position(1, 0), Position(1, 3)),
                    len: 3,
                    content: "abc",
                    level: 0,
                }, None))),
                Some(OperationKind::MathOperation(MathOperation::Sum)),
                Some(Box::new(ExpressionMember::Literal(Token {
                    kind: Some(TokenKind::Int(2)),
                    position: (Position(1, 6), Position(1, 7)),
                    level: 0,
                    content: "2",
                    len: 1
                })))
            ))
        )
        // assert_eq!(parsed.expression.1, Some(OperationKind::MathOperation(MathOperation::Sum)));

        // assert_eq!(
        //     parsed.expression.0,
        //     Token {
        //         content: "abc",
        //         kind: Some(TokenKind::Identifier("abc")),
        //         len: 3,
        //         level: 0,
        //         position: (Position(1, 0), Position(1, 3))
        //     }
        // );

        // assert_eq!(
        //     (parsed.expression.2.as_ref().unwrap())
        //     Token {
        //         content: "2",
        //         kind: Some(TokenKind::Int(2)),
        //         len: 1,
        //         level: 0,
        //         position: (Position(1, 6), Position(1, 7))
        //     }
        // );
    }

    use ra_dev_tools::insta::{assert_json_snapshot, assert_snapshot};
    
    use std::fs::File;
    use std::io::Read;
    use ra_dev_tools::for_each_ra_example_file;

    #[test]
    fn it_should_match_snapshots() {
        println!("{:?}", std::env::args());
        for_each_ra_example_file(|example| {
            let mut file = File::open(example.path()).unwrap();
            
            let mut contents = String::new();
            file.read_to_string(&mut contents).unwrap();

            let block_tree = {
                match parse(&contents) {
                    Ok(b) => b,
                    Err((errors, parsed)) => {

                        let formatted_errors: String = {
                            errors.into_iter().map(|e| format!("{}", e)).collect::<Vec<String>>().join("\n")
                        };

                        
                        assert_snapshot!(String::from(example.path().to_str().unwrap()) + "__ERR__MESSAGES__", formatted_errors);
                        assert_json_snapshot!(String::from(example.path().to_str().unwrap()) + "__ERR__", parsed);


                        panic!("failed to parse example {:?}", example.path());
                    }
                }
            };
            assert_json_snapshot!(example.path().to_str().unwrap(), block_tree)
        })       
    }
}
