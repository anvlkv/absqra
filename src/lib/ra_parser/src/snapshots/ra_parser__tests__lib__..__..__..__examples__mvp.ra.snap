---
source: src/lib/ra_parser/src/tests.rs
expression: block_tree
---
Block {
    kind: Program,
    children: [
        Block {
            kind: Output(
                OutputExpression(
                    Identifier(
                        Token {
                            kind: Some(
                                Identifier(
                                    "parseBlocks",
                                ),
                            ),
                            len: 11,
                            content: "parseBlocks",
                            position: (
                                [3:0],
                                [3:11],
                            ),
                            level: 0,
                        },
                    ),
                    None,
                    None,
                ),
            ),
            children: [
                Block {
                    kind: Input(
                        false,
                        None,
                    ),
                    children: [
                        Block {
                            kind: Invocation(
                                Some(
                                    Token {
                                        kind: Some(
                                            Identifier(
                                                "choice",
                                            ),
                                        ),
                                        len: 6,
                                        content: "choice",
                                        position: (
                                            [5:9],
                                            [5:15],
                                        ),
                                        level: 2,
                                    },
                                ),
                                Some(
                                    InputExpression(
                                        Some(
                                            Named(
                                                Token {
                                                    kind: Some(
                                                        Identifier(
                                                            "multiple",
                                                        ),
                                                    ),
                                                    len: 8,
                                                    content: "multiple",
                                                    position: (
                                                        [5:16],
                                                        [5:24],
                                                    ),
                                                    level: 2,
                                                },
                                            ),
                                        ),
                                        None,
                                        None,
                                    ),
                                ),
                            ),
                            children: [],
                            first_token: Token {
                                kind: Some(
                                    Exclamation,
                                ),
                                len: 1,
                                content: "!",
                                position: (
                                    [5:8],
                                    [5:9],
                                ),
                                level: 2,
                            },
                        },
                        Block {
                            kind: Invocation(
                                Some(
                                    Token {
                                        kind: Some(
                                            Identifier(
                                                "options",
                                            ),
                                        ),
                                        len: 7,
                                        content: "options",
                                        position: (
                                            [6:9],
                                            [6:16],
                                        ),
                                        level: 2,
                                    },
                                ),
                                Some(
                                    InputExpression(
                                        None,
                                        None,
                                        None,
                                    ),
                                ),
                            ),
                            children: [
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "parse blocks",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 14,
                                        content: "parse blocks",
                                        position: (
                                            [7:12],
                                            [7:26],
                                        ),
                                        level: 3,
                                    },
                                },
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "and funny chars too ðŸ˜ ",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 26,
                                        content: "and funny chars too ðŸ˜ ",
                                        position: (
                                            [8:12],
                                            [8:35],
                                        ),
                                        level: 3,
                                    },
                                },
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "â‚°",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 5,
                                        content: "â‚°",
                                        position: (
                                            [9:12],
                                            [9:15],
                                        ),
                                        level: 3,
                                    },
                                },
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "Ï¿",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 4,
                                        content: "Ï¿",
                                        position: (
                                            [10:12],
                                            [10:15],
                                        ),
                                        level: 3,
                                    },
                                },
                            ],
                            first_token: Token {
                                kind: Some(
                                    Exclamation,
                                ),
                                len: 1,
                                content: "!",
                                position: (
                                    [6:8],
                                    [6:9],
                                ),
                                level: 2,
                            },
                        },
                    ],
                    first_token: Token {
                        kind: Some(
                            Greater,
                        ),
                        len: 1,
                        content: ">",
                        position: (
                            [4:4],
                            [4:5],
                        ),
                        level: 1,
                    },
                },
            ],
            first_token: Token {
                kind: Some(
                    Identifier(
                        "parseBlocks",
                    ),
                ),
                len: 11,
                content: "parseBlocks",
                position: (
                    [3:0],
                    [3:11],
                ),
                level: 0,
            },
        },
        Block {
            kind: Output(
                OutputExpression(
                    Identifier(
                        Token {
                            kind: Some(
                                Identifier(
                                    "parseDifferentTypesOfBlocks",
                                ),
                            ),
                            len: 27,
                            content: "parseDifferentTypesOfBlocks",
                            position: (
                                [12:0],
                                [12:27],
                            ),
                            level: 0,
                        },
                    ),
                    None,
                    None,
                ),
            ),
            children: [
                Block {
                    kind: Input(
                        false,
                        None,
                    ),
                    children: [
                        Block {
                            kind: Invocation(
                                Some(
                                    Token {
                                        kind: Some(
                                            Identifier(
                                                "choice",
                                            ),
                                        ),
                                        len: 6,
                                        content: "choice",
                                        position: (
                                            [14:9],
                                            [14:15],
                                        ),
                                        level: 2,
                                    },
                                ),
                                Some(
                                    InputExpression(
                                        Some(
                                            Named(
                                                Token {
                                                    kind: Some(
                                                        Identifier(
                                                            "multiple",
                                                        ),
                                                    ),
                                                    len: 8,
                                                    content: "multiple",
                                                    position: (
                                                        [14:16],
                                                        [14:24],
                                                    ),
                                                    level: 2,
                                                },
                                            ),
                                        ),
                                        None,
                                        None,
                                    ),
                                ),
                            ),
                            children: [],
                            first_token: Token {
                                kind: Some(
                                    Exclamation,
                                ),
                                len: 1,
                                content: "!",
                                position: (
                                    [14:8],
                                    [14:9],
                                ),
                                level: 2,
                            },
                        },
                        Block {
                            kind: Invocation(
                                Some(
                                    Token {
                                        kind: Some(
                                            Identifier(
                                                "options",
                                            ),
                                        ),
                                        len: 7,
                                        content: "options",
                                        position: (
                                            [15:9],
                                            [15:16],
                                        ),
                                        level: 2,
                                    },
                                ),
                                Some(
                                    InputExpression(
                                        None,
                                        None,
                                        None,
                                    ),
                                ),
                            ),
                            children: [
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "comments",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 10,
                                        content: "comments",
                                        position: (
                                            [16:12],
                                            [16:22],
                                        ),
                                        level: 3,
                                    },
                                },
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "var declaration",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 17,
                                        content: "var declaration",
                                        position: (
                                            [17:12],
                                            [17:29],
                                        ),
                                        level: 3,
                                    },
                                },
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "rule declaration",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 18,
                                        content: "rule declaration",
                                        position: (
                                            [18:12],
                                            [18:30],
                                        ),
                                        level: 3,
                                    },
                                },
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "applied rule",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 14,
                                        content: "applied rule",
                                        position: (
                                            [19:12],
                                            [19:26],
                                        ),
                                        level: 3,
                                    },
                                },
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "input block",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 13,
                                        content: "input block",
                                        position: (
                                            [20:12],
                                            [20:25],
                                        ),
                                        level: 3,
                                    },
                                },
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "context switches",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 18,
                                        content: "context switches",
                                        position: (
                                            [21:12],
                                            [21:30],
                                        ),
                                        level: 3,
                                    },
                                },
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "content block",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 15,
                                        content: "content block",
                                        position: (
                                            [22:12],
                                            [22:27],
                                        ),
                                        level: 3,
                                    },
                                },
                            ],
                            first_token: Token {
                                kind: Some(
                                    Exclamation,
                                ),
                                len: 1,
                                content: "!",
                                position: (
                                    [15:8],
                                    [15:9],
                                ),
                                level: 2,
                            },
                        },
                    ],
                    first_token: Token {
                        kind: Some(
                            Greater,
                        ),
                        len: 1,
                        content: ">",
                        position: (
                            [13:4],
                            [13:5],
                        ),
                        level: 1,
                    },
                },
            ],
            first_token: Token {
                kind: Some(
                    Identifier(
                        "parseDifferentTypesOfBlocks",
                    ),
                ),
                len: 27,
                content: "parseDifferentTypesOfBlocks",
                position: (
                    [12:0],
                    [12:27],
                ),
                level: 0,
            },
        },
        Block {
            kind: Output(
                OutputExpression(
                    Identifier(
                        Token {
                            kind: Some(
                                Identifier(
                                    "generateRuntime",
                                ),
                            ),
                            len: 15,
                            content: "generateRuntime",
                            position: (
                                [24:0],
                                [24:15],
                            ),
                            level: 0,
                        },
                    ),
                    None,
                    None,
                ),
            ),
            children: [
                Block {
                    kind: Input(
                        false,
                        None,
                    ),
                    children: [
                        Block {
                            kind: Invocation(
                                Some(
                                    Token {
                                        kind: Some(
                                            Identifier(
                                                "choice",
                                            ),
                                        ),
                                        len: 6,
                                        content: "choice",
                                        position: (
                                            [26:9],
                                            [26:15],
                                        ),
                                        level: 2,
                                    },
                                ),
                                Some(
                                    InputExpression(
                                        Some(
                                            Named(
                                                Token {
                                                    kind: Some(
                                                        Identifier(
                                                            "multiple",
                                                        ),
                                                    ),
                                                    len: 8,
                                                    content: "multiple",
                                                    position: (
                                                        [26:16],
                                                        [26:24],
                                                    ),
                                                    level: 2,
                                                },
                                            ),
                                        ),
                                        None,
                                        None,
                                    ),
                                ),
                            ),
                            children: [],
                            first_token: Token {
                                kind: Some(
                                    Exclamation,
                                ),
                                len: 1,
                                content: "!",
                                position: (
                                    [26:8],
                                    [26:9],
                                ),
                                level: 2,
                            },
                        },
                        Block {
                            kind: Invocation(
                                Some(
                                    Token {
                                        kind: Some(
                                            Identifier(
                                                "options",
                                            ),
                                        ),
                                        len: 7,
                                        content: "options",
                                        position: (
                                            [27:9],
                                            [27:16],
                                        ),
                                        level: 2,
                                    },
                                ),
                                Some(
                                    InputExpression(
                                        None,
                                        None,
                                        None,
                                    ),
                                ),
                            ),
                            children: [
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "render content",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 16,
                                        content: "render content",
                                        position: (
                                            [28:12],
                                            [28:28],
                                        ),
                                        level: 3,
                                    },
                                },
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "render inputs",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 15,
                                        content: "render inputs",
                                        position: (
                                            [29:12],
                                            [29:27],
                                        ),
                                        level: 3,
                                    },
                                },
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "group inputs as long as there no dependency on previous step",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 62,
                                        content: "group inputs as long as there no dependency on previous step",
                                        position: (
                                            [30:12],
                                            [30:74],
                                        ),
                                        level: 3,
                                    },
                                },
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "execute logic",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 15,
                                        content: "execute logic",
                                        position: (
                                            [31:12],
                                            [31:27],
                                        ),
                                        level: 3,
                                    },
                                },
                                Block {
                                    kind: Content(
                                        Content(
                                            Body(
                                                "apply rules",
                                            ),
                                            None,
                                        ),
                                    ),
                                    children: [],
                                    first_token: Token {
                                        kind: Some(
                                            ContentBlock,
                                        ),
                                        len: 13,
                                        content: "apply rules",
                                        position: (
                                            [32:12],
                                            [32:25],
                                        ),
                                        level: 3,
                                    },
                                },
                            ],
                            first_token: Token {
                                kind: Some(
                                    Exclamation,
                                ),
                                len: 1,
                                content: "!",
                                position: (
                                    [27:8],
                                    [27:9],
                                ),
                                level: 2,
                            },
                        },
                    ],
                    first_token: Token {
                        kind: Some(
                            Greater,
                        ),
                        len: 1,
                        content: ">",
                        position: (
                            [25:4],
                            [25:5],
                        ),
                        level: 1,
                    },
                },
            ],
            first_token: Token {
                kind: Some(
                    Identifier(
                        "generateRuntime",
                    ),
                ),
                len: 15,
                content: "generateRuntime",
                position: (
                    [24:0],
                    [24:15],
                ),
                level: 0,
            },
        },
    ],
    first_token: Token {
        kind: None,
        len: 1,
        content: "",
        position: (
            [0:0],
            [0:1],
        ),
        level: 0,
    },
}
